source/slang/slang-type-layout.cpp

```diff
@@ -3613,43 +3661,65 @@ static TypeLayoutResult _createTypeLayout(
+                typeLayout->addResourceUsage(LayoutResourceKind::Uniform, fixedExistentialValueSize + 16);
             }
             typeLayout->addResourceUsage(LayoutResourceKind::ExistentialTypeParam, 1);
             typeLayout->addResourceUsage(LayoutResourceKind::ExistentialObjectParam, 1);
 
             // If there are any concrete types available, the first one will be
             // the value that should be plugged into the slot we just introduced.
             //
-            if( context.specializationArgCount )
+            if (context.specializationArgCount)
             {
                 auto& specializationArg = context.specializationArgs[0];
                 Type* concreteType = as<Type>(specializationArg.val);
                 SLANG_ASSERT(concreteType);
 
-                RefPtr<TypeLayout> concreteTypeLayout = createTypeLayout(context, concreteType);
-
-                // Layout for this specialized interface type then results
-                // in a type layout that tracks both the resource usage of the
-                // interface type itself (just the type + value slots introduced
-                // above), plus a "pending data" type that represents the value
-                // conceptually pointed to by the interface-type field/variable at runtime.
-                //
-                typeLayout->pendingDataTypeLayout = concreteTypeLayout;
+                // Always use AnyValueRules regardless of the enclosing environment's layout rule
+                // for existential values.
+                auto anyValueRules = context.getRulesFamily()->getAnyValueRules();
+
+                // TODO: for traditional GPU targets (HLSL/GLSL) we don't force
+                // anyValueRule for now, since it requires additional work to load
+                // the existential value. We should remove this special case logic
+                // and always use anyValueRule once we implement the correct loading
+                // code gen logic for these targets.
+                if (!(isCPUTarget(context.targetReq) || isCUDATarget(context.targetReq)))
+                    anyValueRules = context.rules;
+
+                RefPtr<TypeLayout> concreteTypeLayout =
+                    createTypeLayout(context.with(anyValueRules), concreteType);
+                auto uniformLayoutInfo =
+                    concreteTypeLayout->FindResourceInfo(LayoutResourceKind::Uniform);
+                if (!useFixedSizedExistentialValue ||
+                    (uniformLayoutInfo && uniformLayoutInfo->count > fixedExistentialValueSize))
+                {
+                    // TODO: for targets that supports pointers, oversized existential values
+                    // should be placed in an overflow region and only a pointer is needed in
+                    // the place of the fixed sized uniform slot.
+                    // We only need the "pending layout" mechanism for targets that does not
+                    // support pointers.
```

Comment:
I suspect that we don't want to use `pendingDataTypeLayout` *at all* for CPU/CUDA.

In the case where things don't fit, we'd have to put a pointer to the actual data into the fixed-size region, and that seems like the only important layout constraint. The code we generate should not care about where the pointed-to bytes reside, so trying to encode that in the layout information seems unnecessary.

I'd like to think of the `pendingDataTypeLayout` thing as *only* being a fix for legacy targets that don't support pointers and have all types be POD-able.